File: web_app/src/core/apiClient.js
import { API_BASE_URL } from './appConfig'
import { tokenStorage } from './tokenStorage'

/// Thin, self-contained failure type every service call throws on non-2xx —
/// screens catch this and show `.message` rather than a raw fetch error.
/// Mirrors mobile_app/lib/core/api_client.dart's ApiException exactly.
export class ApiException extends Error {
  constructor(statusCode, message) {
    super(message)
    this.statusCode = statusCode
    this.message = message
  }
}

/// One shared request function, attaching `Authorization: Bearer <token>`
/// to every request that has a stored session — see
/// context/conventions.md ("API client pattern"). No screen or service
/// ever reads tokenStorage directly to build a header; they call apiRequest
/// (or apiRequestForm, for multipart) and this does it once. Mirrors
/// ApiClient.instance.dio's single-interceptor design in the Flutter app.
async function apiRequest(path, { method = 'GET', body, isForm = false } = {}) {
  const token = tokenStorage.readToken()
  const headers = {}
  if (!isForm) headers['Content-Type'] = 'application/json'
  if (token) headers['Authorization'] = `Bearer ${token}`

  let response
  try {
    response = await fetch(`${API_BASE_URL}${path}`, {
      method,
      headers,
      body: body === undefined ? undefined : isForm ? body : JSON.stringify(body),
    })
  } catch {
    // Covers network failure / connection refused / CORS-style rejection —
    // fetch() throws a plain TypeError for all of these, with no
    // distinguishing detail, same limitation Dio's connectionError branch
    // works around with its own message (see the Flutter equivalent).
    throw new ApiException(null, 'Could not reach RozNoor. Check your connection and try again.')
  }

  let data = null
  const text = await response.text()
  if (text) {
    try {
      data = JSON.parse(text)
    } catch {
      data = null
    }
  }

  if (!response.ok) {
    const message =
      data && typeof data === 'object' && data.detail
        ? String(data.detail)
        : `Request failed (${response.status})`
    throw new ApiException(response.status, message)
  }

  return data
}

export const apiClient = {
  get: (path) => apiRequest(path, { method: 'GET' }),
  post: (path, body) => apiRequest(path, { method: 'POST', body }),
  patch: (path, body) => apiRequest(path, { method: 'PATCH', body }),
  postForm: (path, formData) => apiRequest(path, { method: 'POST', body: formData, isForm: true }),
}

File: mobile_app/lib/core/api_client.dart
import 'package:dio/dio.dart';

import 'app_config.dart';
import 'secure_storage.dart';

/// Thin, self-contained failure type every service call throws on non-2xx —
/// screens catch this and show `message` rather than a raw DioException.
class ApiException implements Exception {
  ApiException(this.statusCode, this.message);

  final int? statusCode;
  final String message;

  @override
  String toString() => message;
}

/// Single Dio instance, shared app-wide, with one interceptor that attaches
/// `Authorization: Bearer <token>` to every request that has a stored
/// session — see context/conventions.md ("API client pattern"). No screen
/// or service ever reads SecureStorageService directly to build a header;
/// they call ApiClient.instance.dio and this interceptor does it once.
class ApiClient {
  ApiClient._() {
    _dio = Dio(
      BaseOptions(
        baseUrl: AppConfig.apiBaseUrl,
        connectTimeout: const Duration(seconds: 15),
        receiveTimeout: const Duration(seconds: 20),
        headers: {'Content-Type': 'application/json'},
      ),
    );
    _dio.interceptors.add(
      InterceptorsWrapper(
        onRequest: (options, handler) async {
          final token = await SecureStorageService.instance.readToken();
          if (token != null) {
            options.headers['Authorization'] = 'Bearer $token';
          }
          handler.next(options);
        },
      ),
    );
  }

  static final ApiClient instance = ApiClient._();

  late final Dio _dio;
  Dio get dio => _dio;

  /// Runs [call] and rethrows any failure as an [ApiException] carrying the
  /// backend's `detail` message (FastAPI's standard error shape) when
  /// present, so every screen gets a human-readable message without
  /// re-parsing DioException itself.
  Future<T> run<T>(Future<T> Function() call) async {
    try {
      return await call();
    } on DioException catch (e) {
      final status = e.response?.statusCode;
      final data = e.response?.data;
      String message;
      if (data is Map && data['detail'] != null) {
        message = data['detail'].toString();
      } else if (e.type == DioExceptionType.connectionTimeout ||
          e.type == DioExceptionType.receiveTimeout ||
          e.type == DioExceptionType.connectionError) {
        message = 'Could not reach RozNoor. Check your connection and try again.';
      } else {
        message = e.message ?? 'Something went wrong.';
      }
      throw ApiException(status, message);
    }
  }
}
